lucet-runtime/lucet-runtime-internals/src/context/mod.rs

```diff
@@ -318,52 +396,52 @@ impl Context {
-
-        // if fptr ever returns, it returns to the backstop func. backstop needs two arguments in
-        // its frame - first the context we are switching *out of* (which is also the one we are
-        // creating right now) and the ctx we switch back into. Note *parent might not be a valid
-        // ctx now, but it should be when this ctx is started.
-        stack[sp - 4] = child as *mut Context as u64;
-        stack[sp - 3] = parent as *mut Context as u64;
-        // Terminate the call chain.
-        stack[sp - 2] = 0;
-        stack[sp - 1] = 0;
+        // set up an initial call stack for guests to bootstrap into and execute
+        let mut stack_builder = CallStackBuilder::new(stack);
+
+        // store arguments we'll pass to `lucet_context_swap` on the stack, above where the guest
+        // might scribble over them.
+        stack_builder.push(parent as *mut Context as u64);
+        stack_builder.push(child as *mut Context as u64);
+
+        // we'll pass a pointer to them via `rbp` in the guest's Context we switch to.
+        let backstop_args = stack_builder.offset();
+
+        // we actually don't want to put an explicit pointer to these arguments anywhere. we'll
+        // line up the rest of the stack such that these are in argument position when we jump to
+        // `fptr`.
+        stack_builder.store_args(spilled_args.as_slice());
+
+        // the stack must be aligned in the environment we'll execute `fptr` from - this is an ABI
+        // requirement and can cause segfaults if not upheld.
+        assert_eq!(
+            stack_builder.offset() % 2,
+            0,
+            "incorrect alignment for guest call frame"
+        );
+
+        // we execute the guest code via returns, so we make a "call stack" of routines like:
+        // -> lucet_context_backstop()
+        //    -> fptr()
+        //       -> lucet_context_bootstrap()
+        //
+        // with each address the start of the named function, so when the inner function
+        // completes it returns to begin the next function up.
+        stack_builder.push(lucet_context_backstop as u64);
+        stack_builder.push(fptr as u64);
+        stack_builder.push(lucet_context_bootstrap as u64);
+
+        let (stack, stack_start) = stack_builder.into_inner();
 
         // RSP, RBP, and sigset still remain to be initialized.
-        // Stack pointer: this has the return address of the first function to be run on the swap.
-        child.gpr.rsp = &mut stack[sp - stack_start] as *mut u64 as u64;
-        // Frame pointer: this is only used by the backstop code. It uses it to locate the ctx and
-        // parent arguments set above.
-        child.gpr.rbp = &mut stack[sp - 2] as *mut u64 as u64;
+        // Stack pointer: this points to the return address that will be used by `swap`, in place
+        // of the original (eg, in the host) return address. This is the address of the first
+        // function to be run on the swap.
```

Comment:
Could this specifically say `lucet_context_bootstrap` to disambiguate from `fptr`?